File: python_programmers/오픈채팅방.py
# 오픈채팅방
# https://programmers.co.kr/learn/courses/30/lessons/42888
# 해결 / 20분 소요 / 21.12.11

from collections import deque
def solution(records):
    answer = []
    user_info = {}
    queue = deque()
    for record in records:
        list_record = record.split()
        # 들어왔을 경우
        if list_record[0] == 'Enter':
            user_info[list_record[1]] = list_record[2]
            queue.append(['Enter', list_record[1]])
        # 나갔을 경우
        elif list_record[0] == 'Leave':
            queue.append(['Leave', list_record[1]])
        # 이름 변경
        else: 
            user_info[list_record[1]] = list_record[2]
    while queue:
        tmp = queue.popleft()
        if tmp[0] == 'Enter':
            answer.append(f"{user_info[tmp[1]]}님이 들어왔습니다.")
        elif tmp[0] == 'Leave':
            answer.append(f"{user_info[tmp[1]]}님이 나갔습니다.")
    return answer

File: javascript_programmers/level2/오픈채팅방.js
// 오픈채팅방
// https://school.programmers.co.kr/learn/courses/30/lessons/42888?language=javascript

// 문제 접근
// id: nickname을 key:value로 만들어두고 변경시 기록해둔다. 그리고, record를 [Enter, Id형식을 기록한다.]


const solution = (record) => {
    const loginIdToNickname = new Map();
    const answers = [];

    record.forEach((value) => {
        const [command, loginId, nickName] = value.split(" ");

        if(command.includes("Leave")){
            answers.push([command, loginId]);
            return;
        }

        if(command.includes("Change")){
            loginIdToNickname.set(loginId, nickName);
            return;
        }

        loginIdToNickname.set(loginId, nickName);
        answers.push([command, loginId]);
    });


    return answers.map(([command, loginId]) => {
        const nickname = loginIdToNickname.get(loginId);
        if(command.includes('Enter') ){
            return `${nickname}님이 들어왔습니다.`;
        }
        return `${nickname}님이 나갔습니다.`;
    });
}